### DISCUSSION TITLE
Using translations 'outside' of React

### DISCUSSION THREAD

**Discussion by PointSingularity** [DoneDeal0/Talkr#28] [Q&A] [answered]
I've enjoyed using talkr a lot, thanks for the great library!

Currently, I am using zod to validate some forms, and some of the errors.
In my usual workflow, I would define the schema outside a React component, but if I want to use talkr it needs to be inside a component/hook.

I can get around this easily by either wrapping a schema inside a memo or creating a new hook:
```tsx
export const LayoutEdit = () => {
    const { T } = useT();

    const layoutEditSchema = useMemo(
        () =>
            z.object({
                Title: z.string().min(1, { message: T('required') }),
                ID: z.string().min(1, { message: T('required') }),
            }),
        [T]
    );
...
```

Out of ergonomic reasons, it would be great to define the schema outside, and use a function in them directly for the error messages.
Is this currently possible to do with talkr?

**Comment by DoneDeal0** [comment, marked as answer]
Hi @PointSingularity , thanks for your message!

I don't know Zod, but you could simply write the translation key in the message field:

```js
    const layoutEditSchema = z.object({
                title: z.string().min(1, { message: "form.required" }),
                id: z.string().min(1, { message: "form.required" ) }),
            })
```

then loop over the errors and write:

```js
<p>{T(error.message)}</p>
```
This way, you can keep your schemas out of your components. Let me know if it works.

**Reply by PointSingularity** [reply]
Hey, thanks for answering! 😄

In theory, it would work, yeah. 
But I stumbled on the whole problem because I had to make the couple of libraries work together like this:

```tsx
                    <NumberInput
                        label={T('customer')}
                        {...form.getInputProps('customer)}
                    />
```

The form returns a series of props, with react node elements, to which I have limited access/control.
The form itself access a schema, where zod comes in.  
So the whole thing is probably more complicated than it should be, that is why it has a lot of limitations in how it needs to be used 🤷‍♂️

Using memo or your approach should work well,
